### PR TITLE
[JNI] Enables fabric handles for CUDA async memory pools

### PR DESCRIPTION
**Pull request by abellina**
Closes https://github.com/rapidsai/cudf/issues/17525
Depends on https://github.com/rapidsai/rmm/pull/1743

## Description
This PR adds a `CUDA_ASYNC_FABRIC` allocation mode in `RmmAllocationMode` and pipes in the options to RMM's `cuda_async_memory_resource` of a `fabric` for the handle type, and `read_write` as the memory protection mode (as that's the only mode supported by the pools, and is required for IPC).

If `CUDA_ASYNC` is used, fabric handles are not requested, and the memory protection is `none`.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by abellina** [comment]
For extra tests: `fabric` is not supported in all versions of CUDA that RAPIDS supports. I can add tests around this, checking for compatibility before, but it would require more code changes.

File: java/src/main/java/ai/rapids/cudf/RmmAllocationMode.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/RmmAllocationMode.java:1]: 2024 copyrights

File: java/src/main/java/ai/rapids/cudf/RmmCudaAsyncMemoryResource.java
Comment (jlowe) [inline java/src/main/java/ai/rapids/cudf/RmmCudaAsyncMemoryResource.java:1]: 2024 copyrights

File: java/src/main/native/src/RmmJni.cpp [hunk 1]
```diff
@@ -775,11 +775,25 @@ JNIEXPORT void JNICALL Java_ai_rapids_cudf_Rmm_releaseArenaMemoryResource(JNIEnv
...
+  
+    // when we are using fabric, we need to set the memory access to be 
+    // read_write, in order for peer GPUs to have access to this memory.
+    auto [handle_type, prot_flag] = !fabric ?
+      std::pair{ 
+        rmm::mr::cuda_async_memory_resource::allocation_handle_type::none,
+        rmm::mr::cuda_async_memory_resource::access_flags::none} :
+      std::pair{
+        rmm::mr::cuda_async_memory_resource::allocation_handle_type::fabric,
+        rmm::mr::cuda_async_memory_resource::access_flags::read_write};
```
Comment (jlowe) [inline java/src/main/native/src/RmmJni.cpp:792]: Nit: It's a little easier to read without the negation
```suggestion
    auto [handle_type, prot_flag] = fabric ?
      std::pair{
        rmm::mr::cuda_async_memory_resource::allocation_handle_type::fabric,
        rmm::mr::cuda_async_memory_resource::access_flags::read_write} :
      std::pair{ 
        rmm::mr::cuda_async_memory_resource::allocation_handle_type::none,
        rmm::mr::cuda_async_memory_resource::access_flags::none};
```

**Review by jlowe — COMMENTED** [review]
copyrights otherwise lgtm

**Review by ttnghia — APPROVED** [review]

**Comment by abellina** [comment]
I saw a failure in existing async tests that was consistent, around "invalid device ordinal". This happened at places where the async allocator would be instantiated. I have changed the code so that, if fabric is not selected, it follows the exact path it used to follow https://github.com/rapidsai/cudf/pull/17526/commits/8d27a927777436c2e2835bf3fce235a9acfaac80. I'll try to repro this, but my guess is the memory access protection APIs don't work on older gpus (this was a V100 in CI).

**Comment by abellina** [comment]
@jlowe fyi

**Review by jlowe — APPROVED** [review]

**Pull request merged**